File: FlutterUI/flutter_ui/lib/core/utilities/service_repository.dart
import 'package:flutter_ui/core/models/delete_model.dart';
import 'package:flutter_ui/core/models/response/list_response_model.dart';
import 'package:flutter_ui/core/models/response/response_model.dart';
import 'package:flutter_ui/core/models/response/single_response_model.dart';

abstract class ServiceRepository<TModel, TId> {
  Future<ResponseModel> add(TModel addModel);

  Future<ResponseModel> delete(DeleteModel deleteModel);

  Future<ResponseModel> update(TModel updateModel);

  Future<ListResponseModel<TModel>> getAll();

  Future<SingleResponseModel<TModel>> getById(TId id);
}

File: AngularUI/angular-ui/src/app/core/services/service-repository.ts
import { ListResponseModel } from '../models/response/listResponseModel';
import { SingleResponseModel } from '../models/response/singleResponseModel';
import { ResponseModel } from '../models/response/responseModel';
import { Observable } from 'rxjs';
import { DeleteModel } from '../models/deleteModel';

export interface ServiceRepository<TModel, TId> {
  add(addModel: TModel): Observable<ResponseModel>;

  delete(deleteModel: DeleteModel): Observable<ResponseModel>;

  update(updateModel: TModel): Observable<ResponseModel>;

  getById(id: TId): Observable<SingleResponseModel<TModel>>;

  getAll(): Observable<ListResponseModel<TModel>>;
}
